File: CS/App_Code/CriteriaVisitor.cs
using DevExpress.Data.Filtering;
using System;
using System.Linq;

public class CriteriaVisitor : CriteriaPatcherBase {
    enum CustomVisitMode { LookingOnlyForCustomOperator, RemoveCustomOperator }
    public static CriteriaOperator RemoveCustomFunction(CriteriaOperator theOperator, string fieldName) {
        return new CriteriaVisitor(fieldName, CustomVisitMode.RemoveCustomOperator).AcceptOperator(theOperator);
    }
    public static string GetCustomFunctionOperatorValue(CriteriaOperator theOperator, string fieldName) {
        var myFunctionOperator = new CriteriaVisitor(fieldName, CustomVisitMode.LookingOnlyForCustomOperator).AcceptOperator(theOperator) as FunctionOperator;
        if(!IsNull(myFunctionOperator))
            return ((ConstantValue)myFunctionOperator.Operands[1]).Value.ToString();
        return null;
    }

    CriteriaVisitor(String fieldName, CustomVisitMode visitMode) {
        FieldName = fieldName;
        VisitMode = visitMode;
    }

    protected string FieldName { get; private set; }
    CustomVisitMode VisitMode { get; set; }

    protected override CriteriaOperator VisitFunction(FunctionOperator theOperator) {
        if(IsMyCustomFunctionOperator(theOperator))
            return VisitMode == CustomVisitMode.LookingOnlyForCustomOperator ? theOperator : null;
        if(VisitMode == CustomVisitMode.RemoveCustomOperator)
            return base.VisitFunction(theOperator);
        return null;
    }
    protected override CriteriaOperator VisitGroup(GroupOperator theOperator) {
        var operands = VisitOperands(theOperator.Operands).Where(op => !IsNull(op));
        if(VisitMode == CustomVisitMode.LookingOnlyForCustomOperator)
            return operands.FirstOrDefault(op => IsMyCustomFunctionOperator(op as FunctionOperator));
        return new GroupOperator(theOperator.OperatorType, operands);
    }
    protected bool IsMyCustomFunctionOperator(FunctionOperator criteria) {
        if(IsNull(criteria) || criteria.OperatorType != FunctionOperatorType.Custom)
            return false;
        var operands = criteria.Operands;
        if(operands.Count != 4)
            return false;
        var customFunctionName = operands[0] as ConstantValue;
        var stateOperand = operands[1] as OperandValue;
        var columnOperand = operands[2] as OperandProperty;
        if(IsNull(customFunctionName) || IsNull(stateOperand) || IsNull(columnOperand))
            return false;
        return customFunctionName.Value.ToString() == MyCustomFunctionOperator.Name && columnOperand.PropertyName == FieldName;
    }
}

File: VB/App_Code/CriteriaVisitor.vb
Option Infer On

Imports DevExpress.Data.Filtering
Imports System
Imports System.Linq

Public Class CriteriaVisitor
    Inherits CriteriaPatcherBase

    Private Enum CustomVisitMode
        LookingOnlyForCustomOperator
        RemoveCustomOperator
    End Enum
    Public Shared Function RemoveCustomFunction(ByVal theOperator As CriteriaOperator, ByVal fieldName As String) As CriteriaOperator
        Return (New CriteriaVisitor(fieldName, CustomVisitMode.RemoveCustomOperator)).AcceptOperator(theOperator)
    End Function
    Public Shared Function GetCustomFunctionOperatorValue(ByVal theOperator As CriteriaOperator, ByVal fieldName As String) As String
        Dim myFunctionOperator = TryCast((New CriteriaVisitor(fieldName, CustomVisitMode.LookingOnlyForCustomOperator)).AcceptOperator(theOperator), FunctionOperator)
        If Not IsNull(myFunctionOperator) Then
            Return CType(myFunctionOperator.Operands(1), ConstantValue).Value.ToString()
        End If
        Return Nothing
    End Function

    Private Sub New(ByVal fieldName As String, ByVal visitMode As CustomVisitMode)
        Me.FieldName = fieldName
        Me.VisitMode = visitMode
    End Sub

    Private privateFieldName As String
    Protected Property FieldName() As String
        Get
            Return privateFieldName
        End Get
        Private Set(ByVal value As String)
            privateFieldName = value
        End Set
    End Property
    Private Property VisitMode() As CustomVisitMode

    Protected Overrides Function VisitFunction(ByVal theOperator As FunctionOperator) As CriteriaOperator
        If IsMyCustomFunctionOperator(theOperator) Then
            Return If(VisitMode = CustomVisitMode.LookingOnlyForCustomOperator, theOperator, Nothing)
        End If
        If VisitMode = CustomVisitMode.RemoveCustomOperator Then
            Return MyBase.VisitFunction(theOperator)
        End If
        Return Nothing
    End Function
    Protected Overrides Function VisitGroup(ByVal theOperator As GroupOperator) As CriteriaOperator
        Dim operands = VisitOperands(theOperator.Operands).Where(Function(op) Not IsNull(op))
        If VisitMode = CustomVisitMode.LookingOnlyForCustomOperator Then
            Return operands.FirstOrDefault(Function(op) IsMyCustomFunctionOperator(TryCast(op, FunctionOperator)))
        End If
        Return New GroupOperator(theOperator.OperatorType, operands)
    End Function
    Protected Function IsMyCustomFunctionOperator(ByVal criteria As FunctionOperator) As Boolean
        If IsNull(criteria) OrElse criteria.OperatorType <> FunctionOperatorType.Custom Then
            Return False
        End If
        Dim operands = criteria.Operands
        If operands.Count <> 4 Then
            Return False
        End If
        Dim customFunctionName = TryCast(operands(0), ConstantValue)
        Dim stateOperand = TryCast(operands(1), OperandValue)
        Dim columnOperand = TryCast(operands(2), OperandProperty)
        If IsNull(customFunctionName) OrElse IsNull(stateOperand) OrElse IsNull(columnOperand) Then
            Return False
        End If
        Return customFunctionName.Value.ToString() = MyCustomFunctionOperator.Name AndAlso columnOperand.PropertyName = FieldName
    End Function
End Class